pkg/storage/batcheval/cmd_add_sstable.go

```diff
@@ -105,6 +113,47 @@ func EvalAddSSTable(
 	}, nil
 }
 
+func checkForKeyCollisions(
+	ctx context.Context,
+	batch engine.ReadWriter,
+	mvccStartKey engine.MVCCKey,
+	mvccEndKey engine.MVCCKey,
+	data []byte,
+) error {
+
+	// Create iterator over the existing data.
+	existingDataIter := batch.NewIterator(engine.IterOptions{UpperBound: mvccEndKey.Key})
+	defer existingDataIter.Close()
+	existingDataIter.Seek(mvccStartKey)
+	if ok, err := existingDataIter.Valid(); err != nil {
+		return errors.Wrap(err, "checking for key collisions")
+	} else if ok && !existingDataIter.UnsafeKey().Less(mvccEndKey) {
+		// Target key range is empty, so it is safe to ingest.
+		return nil
+	}
+
+	// Create iterator over the SST being added.
+	sst := engine.MakeRocksDBSstFileReader()
```

Comment:
this could use a comment explaining that the iteration below is in c++, so we're intentionally firing up a c++ iterator here for it rather than reusing the same (pure go) one we use in verify.

some future commit could potentially also make verify take an iterator as an arg instead of the raw data, and then we could reuse this one if we already had it.